### PR TITLE
[RomApp] Adding intrusive ROM simulation routine via ANN-PROM

### PR DESCRIPTION
**Pull request by NicolasSR**
📝 **Description**
This pull request is the second one in the path to incorporate ANN-PROM into the ROM Application. We already incorporated the training routine for the neural networks in the first PR, and in this one we use these neural networks within the ROM simulation in Kratos. Note that neither petrov-galerkin nor HROM are implemented yet, only global galerkin and global LSPG. Additionally, LineSearch does is not compatible with this type of decoder (this will be solved in the next PR).

Essentially, ANN-PROM is an extension of PROM where a part of the modes form our latent space and are actually manipulated by the solver, and the rest of the modes (or some of them) are estimated by those first ones through a NN to give extra precision and information to the solver.

This implementation is done in a way that will utilize the data already available through the RomManager's database. So, from the user's perspective, they only need to add the NN's configuration within the ROM parameters (as in the last PR), and then set "type_of_decoder" to "ann_enhanced" instead of "linear". Then, any of the typical Fit(), Test(), or RunROM() methods can be called.

Internally, when calling an ANN-PROM simulation, a new global ROM Builder&Solver is called (there is one for galerkin and one for LSPG). All of these use static methods from a new utility called rom_nn_utility.h, which implements in C++ a forward pass of the decoder (including the NN) and the calculation of the derivative of the decoder with respect to its input. The weights of the NN are passed from TensorFlow by the python class NN_ROM_Interface.

**Usage**
The usage is via:
 1. Changing simulation parameter ["ROM"]["type_of_decoder"] to "ann_enhanced".
  2. And then calling any of the following:
- rom_manager.Fit(mu_train, mu_validation): To get the ROM Basis and train the NN.
- rom_manager.Test(mu_train, mu_test): To evaluate an already-existing ROM Basis and NN on a test set. The train set used while training must be specified to retrieve the correct basis and NN weights.
- rom_manager.RunROM(mu_train, mu_run): To run new ROM cases. The train set used while training must be specified to retrieve the correct basis and NN weights.

🆕 **Changelog**
- Added rom_nn_utility.h file: implements RomNNUtility class with various static methods.
- Added ann_prom_global_rom_builder_and_solver.h file: implements AnnPromGlobalROMBuilderAndSolver class, which inherits from ResidualBasedBlockBuilderAndSolver and is an alternative to GlobalROMBuilderAndSolver with different ProjectROM() and SolveROM() methods.
- Added ann_prom_lspg_rom_builder_and_solver.h file: implements AnnPromLeastSquaresPetrovGalerkinROMBuilderAndSolver class, which inherits from AnnPromGlobalROMBuilderAndSolver and is an alternative to LeastSquaresPetrovGalerkinROMBuilderAndSolver with different ProjectROM() method.
- Modified rom_nn_trainer.py file: added the NN_ROM_Interface class, which uses Tensoflow to build the decoder and pass the needed information to the builder&solver.
- Modified rom_manager.py file: modified Fit(), Test() and RunROM() methods to be able to run the ANN-PROM simulations, and to instantiate the NN_ROM_Interface.
- Modified rom_analysis.py file: modified ModifyAfterSolverInitialize() and ModifyInitialGeometry() methods to differentiate between cases with and without ANN-PROM.
- Modified rom_solver.py file: modified CreateSolver class to call the correct Builder&Solvers
- Modified rom_testing_utilities.py file: modified SetUpSimulationInstance() method to be able to pass the NN_ROM_Interface to the analysis stage.

### CODE REVIEW

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 1]
```diff
@@ -644,9 +644,6 @@ class BuilderAndSolver
         return "builder_and_solver";
     }
 
-    ///@}
```
Comment (Rbravo555) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: Let us not touch this file

File: applications/RomApplication/rom_application_variables.h [hunk 2]
```diff
@@ -26,10 +27,26 @@
 
 namespace Kratos
 {
+    // typedef Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> EigenDynamicMatrix;
```
Comment (Rbravo555) [inline applications/RomApplication/rom_application_variables.h]: some cleanup of commented stuff needed

File: applications/RomApplication/rom_application_variables.cpp [hunk 3]
```diff
@@ -20,4 +20,16 @@ namespace Kratos
...
     KRATOS_CREATE_VARIABLE( Vector, ROM_SOLUTION_INCREMENT )
+    KRATOS_CREATE_VARIABLE( Vector, ROM_SOLUTION_BASE )
+    KRATOS_CREATE_VARIABLE( Vector, ROM_SOLUTION_TOTAL )
+    KRATOS_CREATE_VARIABLE( Vector, SOLUTION_BASE )
+    KRATOS_CREATE_VARIABLE( vector<Matrix>, SVD_PHI_MATRICES )
+    KRATOS_CREATE_VARIABLE( vector<Matrix>, NN_LAYERS )
+    KRATOS_CREATE_VARIABLE( Vector, SOLUTION_REFERENCE )
+    KRATOS_CREATE_VARIABLE( bool, UPDATE_PHI_EFFECTIVE_BOOL )
+    
+    // KRATOS_CREATE_VARIABLE( vector<EigenDynamicMatrix>, SVD_PHI_MATRICES )
```
Comment (Rbravo555) [inline applications/RomApplication/rom_application_variables.cpp]: Cleanup, unless required for something. Is it simply a leftover?

**Review by Rbravo555 — COMMENTED** [review]
A more in-depth revision will be provided. But an important comment would be that tests are required for
 - each of the solvers, Galerkin and LSPG
 - for the annprom line search

File: applications/RomApplication/rom_application_variables.cpp (hunk 3 above)
Comment (NicolasSR) [reply applications/RomApplication/rom_application_variables.cpp]: Sure, cleaning it up

**Review by Rbravo555 — APPROVED** [review]
